qa/tasks/cephfs/test_client_recovery.py

```diff
@@ -724,3 +724,11 @@ def test_reconnect_after_blocklisted(self):
 
         rproc.wait()
         self.assertEqual(rproc.exitstatus, 0)
+
+    def test_client_session_reject(self):
+        self.mount_a.umount_wait()
+        self.fs.set_reject_client_session(True)
+        # make sure FUSE client could not be mounted when it's restricted
+        # from establishing a session
+        with self.assertRaises(CommandFailedError):
+            self.mount_a.mount_wait()
```

Comment:
Add a test that:

- keeps client mounted as normal, run a small workload to create a few files (grep around in qa/tasks/cephfs/ for examples)
- brings fs down using `ceph fs fail`
- set the flag reject_cliennt_session
- bring fs back up using `ceph fs set foo joinable true`
- verify the mds complains the client missed reconnecting (grep for `with self.assert_cluster_log`)
- verify the mds has no sessions open

Reply:
This would require me to add a condition in `RECONNECT` case too as it would allow the reconnect to happen cause this patch only blocks the client in case of a new session request

Reply:
@batrick this commit https://github.com/ceph/ceph/pull/48720/commits/fbad4975c86ab4caea6e040a94acbfad63965b57 closes the session if the flag is set and the client tries to reconnect, your thoughts?

Reply:
Looks right.

Reply:
Thanks @batrick. I'll now push the testcase you requested for in sometime now.

Reply:
@batrick extending the testcase, now when I again turn the flag to false i.e. client can establish session now, but the already connected one cannot reconnect as that session was closed, so it is possible to gain it back when the flag is turned false? Or the client now has no option but to start all over? I tested turning flag to false and there are no sessions. In my local machine I can see client is still running:
```
ceph-fuse      fuse.ceph-fuse  107G     0  107G   0% /home/user/ceph/build/mnt
```

but `client ls` and `session ls` are empty, so is it like the `mnt` is useless in this state and either I need another mntpnt or need to umount it first and then start session again?

Reply:
It is possible but let's not complicate the test. Force unmount the client during test cleanup.

Reply:
@batrick testcase added. PTAL. BTW I could not find any relevant cluster log warn/info/debug that would help for the step you mentioned above 
```
verify the mds complains the client missed reconnecting (grep for with self.assert_cluster_log) 
```
closest I could find  was [L1466](https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L1466) and [L1493](https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L1493), and I could not find any of these or any other relevant info in cluster log that would help assert, therefore I added `mds->clog->warn() << "client could not reconnect as<< " file system flag reject_client_session is set";`, check this commit https://github.com/ceph/ceph/pull/48720/commits/41bda6ab7aa32a1b10b952d79eae96913e88bfc6. Not only for this testcase but cluster warning of this kind can be helpful, makes sense?

Reply:
> It is possible but let's not complicate the test. Force unmount the client during test cleanup.

Alright

Reply:
@batrick when the client fails to reconnect and `session_timeout` is reached, the client is evicted and blocklisted. Does this behavior needs to be changed?

Reply:
> Does this behavior needs to be changed here?

No. It's expected.

Reply:
@batrick just thinking out loud, the session timed out because a flag was turned on by us, client had no role in the process, it didn't do anything. Surely we don't want it to have an active session but won't it be too cruel to evict and blocklist it? If not, can you help me understand the rationale behind it?

Reply:
> @batrick when the client fails to reconnect and `session_timeout` is reached, the client is evicted and blocklisted. Does this behavior needs to be changed here?
> 
> I think it happens here https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L1739. I feel instead of evicting and blocklisting we can just kill the session here and move on? What do you think? like:

Do you mean blocklisting a client is way more invasive than abruptly killing the session?

Reply:
> @batrick just thinking out loud, the session timed out because a flag was turned on by us, client had no role in the process, it didn't do anything. Surely we don't want it to have an active session but won't it be too cruel to evict and blocklist it? If not, can you help me understand the rationale behind it?

IMO, this is disabled by default, i.e., clients can (re)connect during recovery. Only is special situations, we'd need the client to stay away.

Reply:
> > @batrick when the client fails to reconnect and `session_timeout` is reached, the client is evicted and blocklisted. Does this behavior needs to be changed here?
> > I think it happens here https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L1739. I feel instead of evicting and blocklisting we can just kill the session here and move on? What do you think? like:
> 
> Do you mean blocklisting a client is way more invasive than abruptly killing the session?

Umm, I don't know. That's what I'm trying to understand here..

Reply:
> > @batrick just thinking out loud, the session timed out because a flag was turned on by us, client had no role in the process, it didn't do anything. Surely we don't want it to have an active session but won't it be too cruel to evict and blocklist it? If not, can you help me understand the rationale behind it?
> 
> IMO, this is disabled by default, i.e., clients can (re)connect during recovery. Only is special situations, we'd need the client to stay away.

Yeah, right but after having this new option, can't we "just keep the client away" instead of evicting and blocklisting it? Is is a better approach?

Reply:
@batrick @vshankar ping?

Reply:
Cool, thanks for the rationale!